functionaltests/8_15_test.go

```diff
@@ -0,0 +1,144 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package functionaltests
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/elastic/apm-server/functionaltests/internal/esclient"
+	"github.com/elastic/apm-server/functionaltests/internal/terraform"
+	"github.com/elastic/go-elasticsearch/v8/typedapi/types"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestUpgrade_8_15_4_to_8_16_0(t *testing.T) {
+	require.NoError(t, ecAPICheck(t))
+
+	start := time.Now()
+	ctx := context.Background()
+
+	t.Log("creating deploment with terraform")
+	tf, err := terraform.New(t, t.Name())
+	require.NoError(t, err)
+	ecTarget := terraform.Var("ec_target", *target)
+	ecRegion := terraform.Var("ec_region", regionFrom(*target))
+	version := terraform.Var("stack_version", "8.15.4")
+	name := terraform.Var("name", t.Name())
+	require.NoError(t, tf.Apply(ctx, ecTarget, ecRegion, version, name))
+	t.Logf("time elapsed: %s", time.Now().Sub(start))
+
+	t.Cleanup(func() {
+		if !t.Failed() || (t.Failed() && *cleanupOnFailure) {
+			t.Log("cleanup terraform resources")
+			require.NoError(t, tf.Destroy(ctx, ecTarget, ecRegion, name, version))
+		} else {
+			t.Log("test failed and cleanup-on-failure is false, skipping cleanup")
+		}
+	})
+
+	var deploymentID string
+	var escfg esclient.Config
+	tf.Output("deployment_id", &deploymentID)
+	tf.Output("apm_url", &escfg.APMServerURL)
+	tf.Output("es_url", &escfg.ElasticsearchURL)
+	tf.Output("username", &escfg.Username)
+	tf.Output("password", &escfg.Password)
+	tf.Output("kb_url", &escfg.KibanaURL)
+
+	t.Logf("created deployment %s", deploymentID)
+
+	ac, err := esclient.New(escfg)
+	require.NoError(t, err)
+
+	t.Log("creating APM API key")
+	apikey, err := ac.CreateAPMAPIKey(ctx, t.Name())
+	require.NoError(t, err)
+
+	ingest(t, escfg.APMServerURL, apikey)
+
+	// Wait few seconds before proceeding to ensure ES indexed all our documents.
+	// Manual tests had failures due to only 4 data streams being reported
+	// when no delay was used. Manual inspection always revealed the correct
+	// number of data streams.
+	time.Sleep(1 * time.Minute)
```

Comment:
I'm not sure I love the 1m wait approach. If we know ahead of time how many documents we are going to end up with, wouldn't it be best to ensure that the aggregate result of `ApmDocCount()` is at least that number? That'll avoid arbitrary waiting times here and there.

Reply:
The upside of this approach was that it had a very simple implementation. I refactored it to verify docs count in some data streams in [`fd73107` (#14935)](https://github.com/elastic/apm-server/pull/14935/commits/fd731071062bc06790269e8492605fa665101098). Only data streams not related to aggregations are checked, as aggregations one have different numbers of documents across runs.

Reply:
I also had to add a check on aggregations data streams, as test where sometimes failing. Added in [`c48e063` (#14935)](https://github.com/elastic/apm-server/pull/14935/commits/c48e0636685469f089d273c46034689cfd95a52f). As the number of document isn't know in advanced I used the docs count not changing as a proxy. In my test this prevented the failure in `1m` aggregation data streams to happen again.